docs/wallet/address-tagging.md

```diff
@@ -0,0 +1,125 @@
+          address: BitcoinAddress,
+          amount: CurrencyUnit,
+          feeRate: FeeUnit,
+          fromAccount: AccountDb,
+          newTags: Vector[AddressTag]) = true
+}
+```
+
+### Using AddressTags
+
+The Bitcoin-S wallet allows you to give addresses, and their associated utxos,
+a tag. These tags allow you to separate funds between utxos so you can query utxos,
+and spend from them, based off of an AddressTag. The system also allows you to create
+your own custom address tags, that will be enforced by the library.
+
+An address tag consists of the tag name, and a tag type. We use a tag type so we can have
+tag with the same name without complications.
+
+To create an address with a tag you can use `getNewAddress` but pass in a `Vector` of `AddressTag`.
+It will add to the address tag database along with all the corresponding tags.
+
+```scala mdoc:silent
+wallet.getNewAddress(tags = Vector(ExampleAddressTag)) 
+```
+
+When sending with `sendToAddress` you can also a `Vector` of new `AddressTag`s that will be applied to the
+resulting change outputs. Any tags of a different tag type not included in `newTag`s will also be applied to
+the change outputs.
+
+```scala mdoc:silent
+wallet.sendToAddress(exampleAddress, Bitcoins(2), SatoshisPerVirtualByte.one, account, Vector(ExampleAddressTag)) 
+```
+
+Also, when sending you can use `fundRawTransaction` and use `fromTagOpt` to pass in an optional `AddressTag`,
+this will use only utxos associated with the `AddressTag`.
+
+```scala mdoc:silent
+wallet.fundRawTransaction(
+    destinations = destinations,
+    feeRate = SatoshisPerVirtualByte.one,
+    fromTagOpt = Some(ExampleAddressTag),
+    markAsReserved = false)
+```
+
+### Creating your own AddressTags
+
+You can create your own custom `AddressTag`s. This allows you to tag addresses and utxos in any way that your
+application needs. To do this you are going to need to use `ExternalAddressTag`.  As an example we will create
+`AddressTag`s for user specific funds.
+
+We will need to define the tag type, then define the tag name for each tag, as well as a way to go to and
+from a `String`. Then we define the actual tags, we are going to have a `Company`, `InsuranceFund`, and `UserId`
+tags. We are going to make the `UserId` tag special, and allow it to take in any user id so we can have a huge
+set of users but all with different ids.
+
+```scala mdoc:silent
+/** Allows to assign funds in a specific address to a user */
+trait UserIdTag extends ExternalAddressTagType {
+  val tagType: String = UserIdTags.tagType
+}
+
+object UserIdTags extends AddressTagFactory[UserIdTag] {
+
+  val tagType: String = "UserIdTag"
+
+  /** Funds that do not belong to any user and instead belong to the company */
+  case object Company extends ExternalAddressTag with UserIdTag {
+    val name: String = "Company"Home
+  }
+
+  /** Funds in the company's insurance fund */
+  case object InsuranceFund extends ExternalAddressTag with UserIdTag {
```

Comment:
`UserIdTag` already extends `ExternalAddressTag` so this isn't needed here or on the others

Reply:
`UserIdTag` extends `ExternalAddressTagType` not `ExternalAddressTag`

Reply:
I think it would be nicer to have `object UserIdTagType extends ExternalAddressTagType` and then `sealed trait UserIdTag extends ExternalAddressTag` and have the tags extend that